File: leetcode/medium/36-validSudoku.js
// Determine if a 9 x 9 Sudoku board is valid. Only the filled cells need to be validated according to the following rules:

// Each row must contain the digits 1-9 without repetition.
// Each column must contain the digits 1-9 without repetition.
// Each of the nine 3 x 3 sub-boxes of the grid must contain the digits 1-9 without repetition.
// Note:

// A Sudoku board (partially filled) could be valid but is not necessarily solvable.
// Only the filled cells need to be validated according to the mentioned rules.

var isValidSudoku = function(board) {
    if (!checkRow(board)) {
        return false;
    }
    
    if (!checkCol(board)) {
        return false;
    }
    
    if (!checkBox(board)) {
        return false;
    }
    return true;
};

function checkRow(board) {
    let regex = /[0-9]/;
    
    for (let row of board) {
        let set = new Set();
        
        for (let col of row) {
            if (regex.test(col) && set.has(col)) {
                return false;
            } else {
                set.add(col)
            }
        }
    }
    return true;
}

function checkCol(board) {
    let regex = /[0-9]/;
    
    for (let i = 0; i < board[0].length; i++) {
        let set = new Set();
        let arr = [board[0][i],board[1][i], board[2][i],board[3][i], board[4][i], 
                  board[5][i], board[6][i], board[7][i], board[8][i]];
        
        for (let num of arr) {
            if (regex.test(num)) {
                if (set.has(num)) return false;
                else set.add(num);
            }  
        }
    }
    return true;
}

function checkBox(board) {
    let regex = /[0-9]/;

    for (let i = 1; i < board.length - 1; i += 3) {
        for (let j = 1; j < board[0].length - 1; j += 3) {
            let set = new Set();
            let arr = [board[i-1][j-1], board[i-1][j], board[i-1][j+1], board[i][j-1], board[i][j], 
                       board[i][j+1], board[i+1][j-1], board[i+1][j], board[i+1][j+1]];
            
            for (let num of arr) {
                if (regex.test(num)) {
                    if (set.has(num)) return false;
                    else set.add(num);
                }
            }
        }
    }
    return true;
}

File: leetcode/medium/36-validSudoku.cpp
class Solution {
public:
    bool isValidSudoku(vector<vector<char>>& board) {
        return checkRows(board) && checkCols(board) && checkSquares(board);
    }

    bool checkRows(vector<vector<char>>& board) {
        for (const auto& row: board) {
                set<int> s;

                for (const auto& col: row) {
                    if (col == '.') continue;
                    if (s.find(col) != s.end()) return false;
                    s.insert(col);
                }
            }
            return true;
    }

    bool checkCols(vector<vector<char>>& board) {
        for (int col = 0; col < board.size(); col++) {
                set<int> s;
                for (int row = 0; row < board[0].size(); row++) {
                    if (board[row][col] == '.') continue;
                    if (s.find(board[row][col]) != s.end()) return false;
                    s.insert(board[row][col]);
                }
            }

        return true;
    }

    bool checkSquares(vector<vector<char>>& board) {
        int startRow = 0;
        int startCol = 0;

        while(startRow < board.size()) {
            set<int> s;

            for (int i = startRow; i < startRow + 3; i++) {
                for (int j = startCol; j < startCol + 3; j++) {
                    if (board[i][j] == '.') continue;
                    if (s.find(board[i][j]) != s.end()) return false;
                    s.insert(board[i][j]);
                }
            }
            if (startCol < 6) {
                startCol += 3;
            } else {
                startRow += 3;
                startCol = 0;
            }
        }

        return true;
    }
};
